src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs

```diff
@@ -19,24 +19,31 @@ internal sealed class SafeFreeCertContext : DebugSafeHandle
     internal sealed class SafeFreeCertContext : SafeHandle
     {
 #endif
-        public SafeFreeCertContext(IntPtr handle) : base(handle, true)
+        private readonly SafeX509Handle _certificate;
+
+        public SafeFreeCertContext(SafeX509Handle certificate) : base(IntPtr.Zero, true)
         {
+            if ((null != certificate) && !certificate.IsInvalid)
+            {
+                _certificate = Interop.libcrypto.X509_dup(certificate);
```

Comment:
I changed the "contract" of SafeFreeCertContext to own the SafeX509Handle passed to it. So instead of X509_dup, it now does an AddRef and also eventually Disposes it.

> The point about needing to know what OpenSSL does still stands, though

Understood. I should qualify my earlier statement as "caller needs to know that OpenSsl (the class) yields ownership of the returned SafeX509Handle but doesn't need to know how it was constructed: